### ISSUE TITLE
RFC: Controlling fp32 matmul’s internal precision

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#76440]
Float32 matmuls are one of the most important operations in PyTorch, and considerable attention has been paid to running them faster with specialized math libraries and new math modes. For example, NVIDIA’s Ampere (and later) CUDA devices support computing fp32 matrix multiplication using tf32 math mode, and Google’s TPUs have [bfloat16](https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus) and bfloat16_3x math modes, all of which perform some of the internal matmul computations in lower precision floating point values that trade numerical accuracy for performance. 

Today in PyTorch these less accurate but faster float32 matmuls are the default, but this has resulted in user confusion (see [this Dev Discuss post](https://dev-discuss.pytorch.org/t/pytorch-and-tensorfloat32/504) summarizing the TensorFloat32 discussion, in particular). While many programs benefit from these faster matmuls, we think it’s better if PyTorch bias towards numerical accuracy by default, and require users explicitly opt-in to these faster math modes when they’re available. We think this is the best way to keep PyTorch users in control and provide them clarity about what their programs are doing.

To enable more generic control over precision of matrix multiplication operation we propose adding a device-agnostic math mode setting, modeled after JAX’s float32 matmul precision UX. This setting would work as follows:

- Add a new function, torch.set_float32_matmul_precision 
  - This function accept values from a FLOAT32_MATMUL_PRECISION enum, as well as strings corresponding to those values
  - The enum has the following values:
    - HIGHEST (corresponding string: “highest”, this is the conceptual default value)
    - HIGH (corresponding string: “high”)
    - MEDIUM (corresponding string: “medium”)
  - The behavior of each flag when passed to the function is as follows:
    - HIGHEST sets torch.backends.cuda.matmul.allow_tf32 = False 
      - Non-native device types should also respect this flag
      - For the XLA device type the corresponding precision setting would be “HIGHEST”
    - HIGH sets torch.backends.cuda.matmul.allow_tf32 = True
      - For the XLA device type the corresponding precision setting would be “HIGH” (bfloat16_3x precision on TPUs)
    - MEDIUM sets torch.backends.cuda.matmul.allow_tf32 = True
      - For the XLA device type the corresponding precision setting would be “DEFAULT” (bfloat16 precision on TPUs)
- The default value of the following flags will be changed:
  - torch.backends.cuda.matmul.allow_tf32 will be changed to False by default
- Non-native device types should change their default float32 matmul mode to be consistent with these settings

(Update: these settings are chosen for consistency with JAX's fp32 matmul precision settings, see [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.matmul.html).)

Note the CPU device type is currently unaffected by this setting.

We think that in the future this mechanism can be used to provide a way for programs to specify their desired float32 matrix multiplication mode across a variety of devices.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @ptrblck @JackCaoG @csarofeen @jeffdaily @Balandat @ngimel

**Comment by stas00** [comment]
> Add a new function, torch.set_float32_matmul_precision

What about getter? May I suggest `torch.float32_matmul_precision` that is a setter when passed an arg and a getter otherwise?

> The enum has the following values:

I'm not at all sure this is an easy to comprehend API since it abstracts away the actual values. e.g. for a normal torch user high and medium mean the same. and highest is sort of meaningless when the user is after an explicit fp32.

It'd be much easier to read the code if it had an explicit:
```
torch.backends.cuda.matmul.allow_tf32 = False
torch.xla = bfloat16_3x # 
```
and what if down the road there will be a new device that has 4 different sub-types?

Perhaps in order to support multiple systems (xla and non-xla in this RFC) instead of abstract enums these could bitwise or like the file open's `O_RDONLY | O_WRONLY`:
```
torch.matmul_precision(MODE_FP32|MODE_TF32|MODE_BF16|MODE_BF16_3x)
```
so future modes from new devices can be supported by adding new flags.

The drawback of this proposal is that older code will have to be changed to support new modes. As compared to the original proposal where the settings will change automatically and require no user code changes.

**Comment by ptrblck** [comment]
I think getter methods might indeed be useful.

CC @eqy could you take a look at this RFC and prepare a PR for it, please?

**Comment by vadimkantorov** [comment]
I also propose allowing per-call optional precision options. In many cases (during performance debugging), this is often more understandable than global options and context managers. In general, it would be nice to improve a quasi-standard mechanism of giving execution hints to ops. Currently, different ops use different naming for this stuff.

Similar issues occur when wanting to force a certain cudnn convolution method to bypass benchmark: https://github.com/pytorch/pytorch/issues/52439

**Comment by jeffdaily** [comment]
Related to #67384, as well?

**Comment by mruberry** [comment]
> Related to #67384, as well?

Yep! This should resolve that RFC, too, this is just a more concrete proposal.

**Comment by vadimkantorov** [comment]
Some functionality where selection of algo/method/precision/approximation used/execution hints is useful and is differently exposed now:
* convs
* cdist
* gelu
* linalg methods (e.g. sqrtm) often possible with different algos having different memory/runtime/stability tradeoffs

**Comment by JackCaoG** [comment]
This proposal sgtm. PyTorch/XLA can implement above methods when they are averrable on the PyTorch core.

**Comment by mruberry** [comment]
Thanks for the comments all! We may want to revisit precision setting more generally in the future (per @vadimkantorov's note) or introduce additional refinements here (per @stas00's thinking), but I think we can best treat those as new issues.

**Comment by vadimkantorov** [comment]
Related discussion: https://github.com/pytorch/pytorch/issues/78581
